ash/src/extensions/khr/maintenance4.rs

```diff
@@ -0,0 +1,98 @@
+use crate::vk;
+use crate::{Device, Instance};
+use std::ffi::CStr;
+use std::mem;
+use std::ops::DerefMut;
+
+#[derive(Clone)]
+pub struct Maintenance4 {
+    handle: vk::Device,
+    fns: vk::KhrMaintenance4Fn,
+}
+
+impl Maintenance4 {
+    pub fn new(instance: &Instance, device: &Device) -> Self {
+        let fns = vk::KhrMaintenance4Fn::load(|name| unsafe {
+            mem::transmute(instance.get_device_proc_addr(device.handle(), name.as_ptr()))
+        });
+        Self {
+            handle: device.handle(),
+            fns,
+        }
+    }
+
+    #[doc = "<https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkGetDeviceBufferMemoryRequirementsKHR.html>"]
+    pub unsafe fn get_device_buffer_memory_requirements<'a>(
+        &self,
+        create_info: &vk::DeviceBufferMemoryRequirementsKHR,
+        next: Option<&'a mut impl vk::ExtendsMemoryRequirements2>,
```

Comment:
Unfortunately I think this sucks to call, because passing `None` (the overwhelmingly common case) leaves the implied type variable ambiguous. I wonder if we could get away with `dyn` instead of `impl` here? Not using (the existing form of) `push_next`, of course, but maybe by hand or with changes to the trait interface?

Reply:
Fair point; `dyn` just works but then again I'm not sure if we shouldn't also keep this consistent with just an `out: &mut vk::MemoryRequirements2` like the slice case.

Though, the idea of this PR is to see if we could/should also change the others. It won't work for the slice case (first commit with the wrong trait bound, before this was force-pushed) since you anyway need to query `_len()` before knowing how many `None`/`Some(&next)` elements to pass in the first place.

Reply:
EDIT: I'm pretty sure I compile-tested this on a different machine to see `impl` fail and `dyn` succeed, but see `error[E0277]: the size for values of type `(dyn definitions::ExtendsMemoryRequirements2 + 'a)` cannot be known at compilation time` now which wasn't an issue before...

Reply:
Right, `push_next` currently requires `Sized` but could be modified not to, or replaced with handwritten code to illustrate the concept.

I think I lean towards mundane consistency here regardless, in part because it's a bit weird to get some of your result data via output argument and some via return value rather than all via output argument. Also keeps things simple. Not a very strong opinion, though.

Reply:
Let's go for simple and consistency then. We can always revisit this in a later PR for the entire API surface at once.